cpp/src/arrow/python/python_to_arrow.cc

```diff
@@ -481,7 +490,7 @@ Status Append(PyObject* context, PyObject* elem, SequenceBuilder* builder,
     }
 #if PY_MAJOR_VERSION < 3
   } else if (PyInt_Check(elem)) {
-    RETURN_NOT_OK(builder->AppendInt64(static_cast<int64_t>(PyInt_AS_LONG(elem))));
+    RETURN_NOT_OK(builder->AppendPy2Int64(static_cast<int64_t>(PyInt_AS_LONG(elem))));
```

Comment:
An issue with this approach is that a serialized `long` with a small numeric value will be deserialized as an `int` which could cause issues for applications relying on type checking.

For example:
```python
In [9]: value = 0L

In [10]: type(value)
Out[10]: long

In [11]: serialized = pa.serialize(value)

In [12]: deserialized = serialized.deserialize()

In [13]: type(deserialized)
Out[13]: int
```

Reply:
Accepting only `long` and refusing `int` sounds like a weird thing to do, though. Can you point an API that works this way?
(there are rare APIs that do the reverse - accept only `int` and reject `long` - but usually it's because they were written in C and the programmer couldn't care more)

Reply:
Sorry, I don't know one off the top of my head -- I agree that it's bad practice. I do think that being consistent with types is important from a correctness point of view.
Pickle, for reference, does not deserialize a small `long` as an `int`.

Reply:
That's true. Let's go with this patch, then. Hopefully we'll ditch Python 2 compatibility soon ;-)